// @alphafounders/rf-questionflow-component/src/components/questions/RichMultiChoice/RichMultiChoice.stories.tsx
import React from 'react';
import { Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { insurerBanksData } from '../../QuestionFlow.mock';
import RichMultiChoice, { RichMultiChoiceProps } from './RichMultiChoice';

export default {
  component: RichMultiChoice,
  title: 'questions/RichMultiChoice',
};

const Template: Story<RichMultiChoiceProps> = (args) => (
  <RichMultiChoice {...args} />
);

const formikValues = {
  getFieldProps: () => action('get field props'),
  values: {
    rich_multi_choice: [3],
  },
  touched: {},
  errors: {},
};

const question = {
  code: 'rich_multi_choice',
  label: 'Rich Multi Choice Label',
  limitationText: (
    <span
      style={{
        color: '#005098',
        fontSize: '1rem',
      }}
    >
      * Thai citizenship only
    </span>
  ),
  description: {
    text: 'Rich Multi Choice Description',
    tip: 'Rich Multi Choice Tip Text',
  },
  checkbox: {
    label: 'Let Rabbit Care choose for you?',
    value: '1',
  },
  items: insurerBanksData,
  subLabel: (
    <div>
      Current Value of your car :
      <b style={{ color: '#035098', marginRight: '0.4rem' }}> 500,000 THB</b>
      <br />
      (reference Price)
    </div>
  ),
};

export const Default = Template.bind({});
Default.args = {
  formik: formikValues,
  t: (string) => string,
  question,
  handleNextButton: () => action('on next button'),
  handleInputChange: () => action('on input change'),
};

export const Recommended = Template.bind({});
Recommended.args = {
  formik: formikValues,
  t: (string) => string,
  question,
  recommendedItems: [1, 2, 3],
  handleNextButton: () => action('on next button'),
  handleInputChange: (fieldName, list) => {
    formikValues.values[fieldName] = list;
  },
};
