// @alphafounders/rf-questionflow-component/src/components/QuestionFlow.mock.tsx
export const insurerBanksData = [
  {
    id: 'SCB',
    name: 'SCB',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: true,
    isActive: true,
    details: [
      {
        label: 'Interest Rate',
        value: '~ 5.05% flat rate (= 10.58% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 5M',
      },
      {
        label: 'Max. term loan',
        value: '84 months',
      },
      {
        label: 'NCB check',
        value: 'Yes',
      },
    ],
  },
  {
    id: 'KKP',
    name: 'Kiatnakin Phatra',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: false,
    isActive: true,
    details: [
      {
        label: 'Interest Rate',
        value: '~ 6.65% flat rate (= 11.97% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 5M',
      },
      {
        label: 'Max. term loan',
        value: '84 months',
      },
      {
        label: 'NCB check',
        value: 'Yes',
      },
    ],
  },
  {
    id: 'KASIKORN',
    name: 'Kasikorn',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: false,
    isActive: true,
    details: [
      {
        label: 'Interest Rate',
        value: '~ 4.06% flat rate (= 7.31% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 5M',
      },
      {
        label: 'Max. term loan',
        value: '72 months',
      },
      {
        label: 'NCB check',
        value: 'Yes',
      },
    ],
  },
  {
    id: 'KRUNGSRI',
    name: 'Krungsri',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: false,
    isActive: true,
    details: [
      {
        label: 'Interest Rate',
        value: '~ 3.09% flat rate (= 5.57% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 5M',
      },
      {
        label: 'Max. term loan',
        value: '84 months',
      },
      {
        label: 'NCB check',
        value: 'Yes',
      },
    ],
  },
  {
    id: 'TISCO',
    name: 'TISCO',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: false,
    isActive: true,
    details: [
      {
        label: 'Interest Rate',
        value: '~ 5.88% flat rate (= 10.58% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 10M',
      },
      {
        label: 'Max. term loan',
        value: '72 months',
      },
      {
        label: 'NCB check',
        value: 'Yes',
      },
    ],
  },
  {
    id: 'ASN',
    name: 'ASN',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: false,
    isActive: true,
    details: [
      {
        label: 'Interest Rate',
        value: '~ 9.48% flat rate (= 17.06% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 300K',
      },
      {
        label: 'Max. term loan',
        value: '48 months',
      },
      {
        label: 'NCB check',
        value: 'No',
      },
    ],
  },
  {
    id: 'SOMWANG',
    name: 'Somwang',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: false,
    isActive: false,
    eligibleText: 'You are not eligible for this bank',
    details: [
      {
        label: 'Interest Rate',
        value: '~ 5.88% flat rate (= 10.58% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 10M',
      },
      {
        label: 'Max. term loan',
        value: '72 months',
      },
      {
        label: 'NCB check',
        value: 'No',
      },
    ],
  },
  {
    id: 'CARFINN',
    name: 'CarFinn',
    logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    isRecommended: false,
    isActive: false,
    eligibleText: 'You are not eligible for this bank',
    details: [
      {
        label: 'Interest Rate',
        value: '~ 3.66% flat rate (= 6.48% effective)',
      },
      {
        label: 'Max. loan amount',
        value: 'Car value but not exceeding THB 300K',
      },
      {
        label: 'Max. term loan',
        value: '72 months',
      },
      {
        label: 'NCB check',
        value: 'No',
      },
    ],
  },
];
