// @alphafounders/rf-questionflow-component/src/components/questions/Price.stories.tsx
import React from 'react';
import { Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Price, { PriceProps } from './Price';

export default {
  component: Price,
  title: 'questions/Price',
};

const Template: Story<PriceProps> = (args) => <Price {...args} />;

const formikValues = {
  getFieldProps: () => action('get field props'),
  values: {},
  touched: {},
  errors: {},
};

const conditionalFormikValues = {
  getFieldProps: () => action('get field props'),
  values: {},
  touched: {
    car_loan_price: true,
  },
  errors: {
    car_loan_price: 'Required',
  },
};

const validFormikValues = {
  getFieldProps: () => action('get field props'),
  values: {
    car_loan_price: 10000000,
  },
  touched: {
    car_loan_price: true,
  },
  errors: {
    car_loan_price: '',
  },
};

const question = {
  code: 'car_loan_price',
  label: 'Car Loan Label',
  currency: 'THB',
  showIcon: true,
  subLabel: (
    <div>
      Current Value of your car :
      <b style={{ color: '#035098', marginRight: '0.4rem' }}> 500,000 THB</b>
      <br />
      (reference Price)
    </div>
  ),
};

export const Default = Template.bind({});
Default.args = {
  formik: formikValues,
  t: (string) => string,
  question,
  handleNextButton: () => action('on next button'),
  handleInputChange: () => action('on input change'),
};

export const Conditional = Template.bind({});
Conditional.args = {
  formik: conditionalFormikValues,
  t: (string) => string,
  question,
  handleNextButton: () => action('on next button'),
  handleInputChange: () => action('on input change'),
};

export const Valid = Template.bind({});
Valid.args = {
  formik: validFormikValues,
  t: (string) => string,
  question,
  handleNextButton: () => action('on next button'),
  handleInputChange: () => action('on input change'),
};
