// @alphafounders/rf-questionflow-component/src/components/questions/MultiSelectField/Item.tsx
import React, { ReactElement } from 'react';
import './MultiSelectField.modules.scss';
import CheckIcon from '../../../public/icons/common/checked.svg';
import UnCheckIcon from '../../../public/icons/common/unchecked.svg';

export type IItem = {
  id: string;
  name: string;
  logo: string;
  isReset?: boolean;
};

export type ItemProps = {
  t: (s: string) => string;
  item: IItem;
  handleInputChange: (value: string, status: boolean | undefined) => void;
  isChecked?: boolean | undefined;
};

const Item = ({
  t,
  item,
  handleInputChange,
  isChecked,
}: ItemProps): ReactElement => {
  const onClickCard = () => {
    handleInputChange(item?.id, isChecked);
  };

  return (
    <div className={`${'multi-select__card-section'}`} onClick={onClickCard}>
      <div
        data-testid={item.id}
        id={`${item.id}__detail-card`}
        className={`card multi-select__detail-card ${
          isChecked ? 'multi-select__selected-item' : ''
        }`}
      >
        <div className="card-body">
          <div className="row multi-select__logo-row">
            <div className="col-10">
              <img
                data-testid="multi-select-logo"
                className="img-fluid"
                srcSet={item.logo}
                alt="Multi Select Logo"
                style={{ height: '32px' }}
              />
              <h6 className="d-inline">{t(item.name)}</h6>
            </div>
            <div className="col-2 my-auto">
              <div>
                {isChecked ? (
                  <CheckIcon style={{ float: 'right' }} />
                ) : (
                  <UnCheckIcon style={{ float: 'right' }} />
                )}
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Item;
