// @alphafounders/rf-questionflow-component/src/components/questions/MultiSelectField/MultiSelectField.stories.tsx
import React from 'react';
import { Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import MultiSelectField, { MultiSelectFieldProps } from './MultiSelectField';

export default {
  component: MultiSelectField,
  title: 'questions/MultiSelectField',
};

const Template: Story<MultiSelectFieldProps> = (args) => (
  <MultiSelectField {...args} />
);

const formikValues = {
  getFieldProps: () => action('get field props'),
  values: {
    multi_select_field: ['3'],
  },
  touched: {},
  errors: {},
};

const question = {
  code: 'multi_select_field',
  label: 'Multi Select Field Label',
  items: [
    {
      id: '1',
      name: 'SCB',
      logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    },
    {
      id: '2',
      name: 'Kiatnakin Phatra',
      logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    },
    {
      id: '3',
      name: 'Kasikorn',
      logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    },
    {
      id: '4',
      name: 'Krungsri',
      logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    },
    {
      id: '5',
      name: 'TISCO',
      logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    },
    {
      id: '6',
      name: 'ASN',
      logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
    },
    {
      id: '7',
      name: 'Somwang',
      logo: 'https://staging-finance.rabbitinternet.com/qflow/static/insurers/7.png?0 1x, https://staging-finance.rabbitinternet.com/qflow/static/insurers/7@2x.png?0 2x',
      isReset: true,
    },
  ],
};

export const Default = Template.bind({});
Default.args = {
  formik: formikValues,
  t: (string) => string,
  question,
  handleNextButton: () => action('on next button'),
  handleInputChange: (fieldName, list) => {
    formikValues.values[fieldName] = list;
  },
};
