// @alphafounders/rf-questionflow-component/src/components/questions/SquareCheckboxes/SquareCheckboxes.stories.tsx
import React from 'react';
import { Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import SquareCheckboxes, { SquareCheckboxesProps } from '.';

export default {
  component: SquareCheckboxes,
  title: 'questions/SquareCheckboxes',
};

const Template: Story<SquareCheckboxesProps> = (args) => (
  <SquareCheckboxes {...args} />
);

const formikValues = {
  getFieldProps: () => action('get field props'),
  values: {},
  touched: {},
  errors: {},
};

const checkboxesValues = [
  {
    value: 1,
    label: 'Type 1',
    title: 'Full Coverage of vehicle, medical and 3rd party claims',
    desc: 'Standard plan with full coverage for life and injury and property of third parties. Has fire and theft coverage and insure drivers property damage and liability',
    iconTitle: 'icon',
  },
  {
    value: 2,
    label: 'Type 2',
    title: 'Limited Coverage of vehicle, medical and 3rd party claims',
    desc: 'Covering life and injury and property damage of third party. Insure in case of theft and fire. Insure drivers property damage (only vehicle collision)',
    iconTitle: 'icon',
  },
  {
    value: 3,
    label: 'Type 3',
    title: 'Limited Coverage of vehicle, medical and 3rd party claims',
    desc: 'Covering life and injury and property damage of third party. Insure in case of theft and fire. Insure drivers property damage (only vehicle collision)',
    iconTitle: 'icon',
    disabled: true,
  },
];

export const Default = Template.bind({});
Default.args = {
  formik: formikValues,
  t: (string) => string,
  checkboxes: checkboxesValues,
  fieldName: 'square-checkbox',
  tip: 'square-tip',
  label: 'motor:questions.labels.square_checkbox',
  handleNextButton: () => action('on next button'),
  onSubmit: () => action('on submit'),
};
