// @alphafounders/rf-questionflow-component/src/components/common/Modal.tsx
import ReactModal from 'react-modal';
import React, { useEffect, ReactElement } from 'react';
import './Modal.module.scss';

export type ModalProps = {
  isOpen: boolean;
  onClose?: () => void;
  children: ReactElement;
};

const Modal = ({ isOpen, onClose, children }: ModalProps): ReactElement => {
  useEffect(() => {
    ReactModal.setAppElement('body');
  }, []);

  useEffect(() => {
    if (isOpen) {
      document.body.style.overflow = 'hidden';
    }

    return () => {
      document.body.style.overflow = 'scroll';
    };
  }, [isOpen]);

  return (
    <ReactModal
      isOpen={isOpen}
      onRequestClose={onClose}
      portalClassName={`${['rf-modal']}`}
      overlayClassName={`${['rf-modal__overlay']}`}
      className={`${['rf-modal__content']}`}
      preventScroll
      style={{
        overlay: {
          zIndex: 1076,
          backgroundColor: 'rgba(8, 31, 67, 0.9)',
        },
      }}
    >
      <div className={`${['rf-modal__content__wrapper']}`}>
        <div className={`${['rf-modal__content__wrapper__main-container']}`}>
          {children}
        </div>
      </div>
    </ReactModal>
  );
};

export default Modal;
